src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,683 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace ILLink.RoslynAnalyzer
+{
+	[DiagnosticAnalyzer (LanguageNames.CSharp)]
+	public class DynamicallyAccessedMembersAnalyzer : DiagnosticAnalyzer
+	{
+		internal const string DynamicallyAccessedMembers = nameof (DynamicallyAccessedMembers);
+		internal const string DynamicallyAccessedMembersAttribute = nameof (DynamicallyAccessedMembersAttribute);
+
+		#region SupportedDiagnostics
+		public const string IL2067 = "IL2067";
+		static readonly DiagnosticDescriptor s_IL2067 = new (
+			IL2067,
```

Comment:
Rant: This feels like a great opportunity for some kind of source generator - having to write this much code is crazy.

For now - can we refactor this a bit - create a helper which only takes the parts which are different and call that instead of repeating the same parameters over and over.